### PR TITLE
Engine config tests

### PR DESCRIPTION
**Pull request by ditwoo**
## Before submitting (checklist)

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [ ] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [ ] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [ ] Did you write any new necessary tests?
- [ ] Did you check that your code passes the unit tests `pytest .` ? 
- [ ] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

PS
- [x] I know, that I could [join slack](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for pull request discussion.

### CODE REVIEW

File: catalyst/callbacks/checkpoint.py [hunk 1]
```diff
@@ -568,11 +568,13 @@ def on_epoch_end(self, runner: "IRunner") -> None:
                 dict(runner.epoch_metrics),
             )
             self.top_best_metrics.append(metrics_record)
-            # truncate checkpoints
-            self._truncate_checkpoints()
-            # save checkpoint metrics
-            metrics_log = self._prepare_metrics_log(float(score), dict(runner.epoch_metrics))
-            save_config(metrics_log, f"{self.logdir}/{self.metrics_filename}")
+            # do operations only from master process
+            if runner.engine.is_master_process:
```
Comment (Scitator) [inline catalyst/callbacks/checkpoint.py:572]: Checkpoint callback should be `master`-node only, isn't it?

File: catalyst/engines/__init__.py [hunk 2]
```diff
@@ -13,9 +13,21 @@
 
 if SETTINGS.amp_required:
     from catalyst.engines.amp import AMPEngine, DistributedDataParallelAMPEngine
+else:
+    # replacement
+    from catalyst.engines.device import DeviceEngine as AMPEngine
+    from catalyst.engines.distributed import (
+        DistributedDataParallelEngine as DistributedDataParallelAMPEngine,
+    )
```
Comment (Scitator) [inline catalyst/engines/__init__.py]: why do we need it? let's do not mess the instances of the engines
it could become unreproducible in this case

File: catalyst/engines/__init__.py [hunk 3]
```diff
@@ -76,3 +116,9 @@ def process_engine(engine: Union[str, IEngine, None] = None) -> IEngine:
     "DistributedDataParallelEngine",
     "process_engine",
 ]
+
+if SETTINGS.amp_required:
+    __all__ += ["AMPEngine", "DistributedDataParallelAMPEngine"]
+
+if SETTINGS.apex_required:
+    __all__ += ["APEXEngine", "DistributedDataParallelApexEngine"]
```
Comment (Scitator) [inline catalyst/engines/__init__.py:33]: :feelsgood:

File: catalyst/core/engine.py [hunk 4]
```diff
@@ -81,7 +93,12 @@ def optimizer_step(self, loss, model, optimizer) -> None:
 
     @abstractmethod
     def pack_checkpoint(
-        self, model=None, criterion=None, optimizer=None, scheduler=None, **kwargs,
+        self,
+        model: Model = None,
+        criterion: Criterion = None,
+        optimizer: Optimizer = None,
+        scheduler: Scheduler = None,
```
Comment (Scitator) [inline catalyst/core/engine.py:100]: should they be `RunnerModel`, `RunnerCriterion`, etc? from `catalyst.typing`

File: catalyst/runners/config.py [hunk 5]
```diff
@@ -186,8 +186,8 @@ def _get_criterion_from_params(**params) -> RunnerCriterion:
 
         if key_value_flag:
             criterion = {
-                key: ConfigRunner._get_criterion_from_params(**key_params)  # noqa: WPS437
-                for key, key_params in params.items()
+                key: ConfigRunner.get_criterion_(**key_params)
+                for key, key_params in params.items()  # noqa: WPS437
```
Comment (Scitator) [inline catalyst/runners/config.py]:
```suggestion
                key: ConfigRunner._get_criterion_from_params(**key_params)  # noqa: WPS437
                for key, key_params in params.items()

```

File: catalyst/runners/config.py [hunk 6]
```diff
@@ -281,9 +281,9 @@ def _get_scheduler_from_params(*, optimizer: RunnerOptimizer, **params) -> Runne
         if is_key_value:
             scheduler: Dict[str, Scheduler] = {}
             for key, scheduler_params in params.items():
-                scheduler[key] = ConfigRunner._get_scheduler_from_params(  # noqa: WPS437
+                scheduler[key] = ConfigRunner.get_scheduler_(
                     **scheduler_params, optimizer=optimizer
-                )
+                )  # noqa: WPS437
```
Comment (Scitator) [inline catalyst/runners/config.py]: return the `_get_scheduler_from_params` please

**Review by Scitator — COMMENTED** [review]
👍

File: catalyst/callbacks/checkpoint.py (hunk 1 above)
Comment (ditwoo) [reply catalyst/callbacks/checkpoint.py:572]: Yeah

File: catalyst/engines/__init__.py (hunk 2 above)
Comment (ditwoo) [reply catalyst/engines/__init__.py]: Sure, I'll remove "replacement" imports from engines

File: catalyst/core/engine.py (hunk 4 above)
Comment (ditwoo) [reply catalyst/core/engine.py:100]: I think that Model, Criterion, Optimizer, and Scheduler is a base blocks and engine should work only with pytorch primitives.

Comment (Scitator) [reply catalyst/core/engine.py:100]: why shouldn't it also work with `Dict[Model]` case? that's what `RunnerModel` is about

File: catalyst/callbacks/checkpoint.py (hunk 1 above)
Comment (Scitator) [reply catalyst/callbacks/checkpoint.py:572]: I mean, that the whole callback would be executed only on the master process, why do we need extra ifs here?

File: catalyst/core/engine.py (hunk 4 above)
Comment (Scitator) [reply catalyst/core/engine.py:100]: :top:

File: catalyst/engines/tests/test_metrics.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+    def stages(self):
+        return ["train"]
+
+    def get_stage_len(self, stage: str) -> int:
+        return 10
+
+    def get_loaders(self, stage: str, epoch: int = None) -> Dict[str, Any]:
+        dataset = TwoBlobsDataset()
+        loader = DataLoader(dataset, batch_size=_BATCH_SIZE, num_workers=_WORKERS)
+        return {"train": loader, "valid": loader}
```
Comment (Scitator) [inline catalyst/engines/tests/test_metrics.py:124]: do we need `train` loader here? we could just compare metrics on the `valid` one without any dataset/model changes
what do you think about it?

**Comment by Scitator** [comment]
merged during https://github.com/catalyst-team/catalyst/pull/1100 :)

PS. let's use catalyst-based branches for shared development please
